design/high-level-reactivity.md

```diff
@@ -0,0 +1,236 @@
+        _deps: {
+          _self: { key: '12341234', version: 3 },
+          tasks: { key: '35232345', version: 4 }
+        }
+      }
+    ]
+  }
+}
+```
+
+This will tell the client which dependencies they should watch to figure out if the list object itself changed, or if the list of tasks needs to be updated.
+
+Note that can’t put the `_deps` field on tasks itself, because JSON doesn’t allow it, so we have to put it on the parent instead. Also, the `_self` field is a shorthand to avoid listing all of the properties of the list object - it could have `name`, `description`, and so on, and it would be a waste of bandwidth to send all of those key names again.
+
+### Automatically recording dependencies when reading data
+In order to know when a GraphQL query needs to be re-run, we need to know what deps represent different parts of the query. These deps can be recorded manually for complex queries, but simple queries can have their deps identified automatically. For example, here’s a query used in a particular part of the GraphQL resolution tree:
+
+```
+todoLists.select('*').where('id', 1);
+```
+
+This might record the following dependency:
+
+```
+{ key: 'todoLists:1', version: 0 }
+```
+
+The dependency recording mechanism queries the dependency tracking server to figure out the current version.
+
+### Manually recording dependencies
+
+The automatic dependency recording mechanism won’t work if it isn’t possible to analyze the query to determine the dependencies. In these cases, the developer will need to manually record a dependency, using any string key they want.
+
+For example, if you have a complex query to count the number of notifications a user has, you might want a custom invalidation key for that number:
+
+```
+// somewhere, a function that consistently generates a key
+function notificationDepKeyForUser(userId) {
+  return 'notificationCount:' + userId;
+}
+
+// inside the GraphQL resolver
+numNotifications = getNotificationCountForUser(userId);
+context.recordDependency(notificationDepKeyForUser(userId));
+```
+
+This will let you manually specify situations in which the number of notifications should be refreshed. Advanced users with particularly strict performance requirements, like a site-wide visitor count or a realtime high-score table, might also choose to use manually constructed dependencies for more control over their data flow.
+
+## Simple reactivity model
+
+Based on the above primitives, here is a stateless strategy for GraphQL reactivity:
+
+1. The client fetches the query from the GraphQL server, which includes a set of deps in the response.
+2. The client periodically polls the invalidation server with the set of deps, and the server returns the list of deps which have a newer version available. Note that this can easily be converted to a stateful approach where the client subscribes to a list of depenencies over a websocket, for parts of the client which need lower latency.
```

Comment:
What if the websocket path would only be used to push if a certain "dep id" has been changed. Then using that as the trigger for the client to refetch all that have these "dep ids"?

I never cared much for the "per field reactivity" as much, but I did like the principle of not having to poll. Thinking about it, if GraphQL is a dynamic per field like subscription, isn't the above approach than not always enough? E.g., this wouldn't have a downside regarding needing to fetch unwanted data does it? Since that is what the GraphQL definition decides (and therefore limits).